spinalcordtoolbox/reports/qc.py

```diff
@@ -660,7 +653,13 @@ def _update_html_assets(self, json_data):
         with io.open(os.path.join(assets_path, 'index.html'), encoding="utf-8") as template_index:
             template = Template(template_index.read())
             output = template.substitute(sct_json_data=json.dumps(json_data))
-            io.open(os.path.join(dest_path, 'index.html'), 'w', encoding="utf-8").write(output)
+
+        # Use locking when writing to `index.html` only (since this is the only file that multiple processes
+        # would simultaneously write to)
+        dest_file = io.open(os.path.join(dest_path, 'index.html'), 'w', encoding="utf-8")
+        portalocker.lock(dest_file, portalocker.LOCK_EX)
```

Comment:
I'm seeing some possible race conditions with the current ordering. The biggest one would be:
1. Process A starts writing to A.json
2. Process B reads all json files, including this probably-malformed-right-now A.json, and crashes.

Another would be:
1. Process C creates C.json, and reads all json files.
2. Process D creates D.json, reads all json files, produces a new index.html, with all the data.
3. Process C overwrites this with a new index.html file, which doesn't include D.json.

With portalocker, my understanding is that we can only lock files, not directories, but that's not a problem: we could lock any particular file, like index.html, as long as all the processes use the same file. But the ordering should probably be something like:
1. Open index.html for updating, mode `'r+'` (not writing! That would immediately delete the contents), and grab/wait for an exclusive lock.
2. Write a new json file.
3. Read all json files.
4. Write the new index.html.
5. Release the lock on index.html.
6. (I'm not sure if this step needs to hold the lock, since these files are constant) copy the other asset files.

That way, all the modifications to (the collection of json files and their contents) happen within the critical region, so each process has a consistent view, and the last update to index.html contains everything.

Reply:
> 2. Process B reads all json files, including this probably-malformed-right-now A.json, and crashes.

My goodness... I think you may have predicted the [exact error](https://github.com/spinalcordtoolbox/spinalcordtoolbox/runs/5379684935?check_suite_focus=true) that happened to me earlier today?

```
________________________________ test_many_qc _________________________________
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\testing\api\test_qc_parallel.py", line 15, in gen_qc
    qc.generate_qc(fname_in1=t2_image, fname_seg=t2_seg, path_qc=path_qc, process="sct_deepseg_gm")
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\reports\qc.py", line 878, in generate_qc
    fps=fps
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\reports\qc.py", line 717, in add_entry
    layout(qcslice)
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\reports\qc.py", line 317, in wrapped_f
    self._make_QC_image_for_3d_volumes(img, mask, slice_orientation=sct_slice.get_name())
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\reports\qc.py", line 364, in _make_QC_image_for_3d_volumes
    self.qc_report.update_description_file(img.shape)
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\reports\qc.py", line 646, in update_description_file
    self._update_html_assets(get_json_data_from_path(path_json))
  File "d:\a\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\spinalcordtoolbox\reports\qc.py", line 888, in get_json_data_from_path
    results.append(json.load(fjson))
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\json\__init__.py", line 296, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\json\__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\hostedtoolcache\windows\Python\3.7.9\x64\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
"""
```

When writing this original `portalocker` fix, I thought that the `index.html` file was the only one that was repeatedly read and written to. But, as you've identified, the `.json` files are read and written to as well. Good catch!! Pat yourself on the back for that, gosh.

Your proposed steps make sense to me, too. Locking first before touching any `.json`/`.html` files does seem to me like it will prevent any issues with concurrent reads/writes of `.json` files.

Reply:
Done in [f2c2453](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3701/commits/f2c24536d0c57d04c7fb286673fb4dc42d7f43da).